lttoolbox/transducer.cc

```diff
@@ -1395,3 +1395,204 @@ Transducer::applyACX(Alphabet& alpha,
+   * where f = this
+   *
+   * The basic algorithm is:
+   *
+   * Transducer gf;             // composition
+   * queue = [ (f.init, g.init, gf.init) ]
+   * while(!queue.empty()):
+   *   f_src, g_src, gf_src = queue.pop()
+   *   for (f_input, f_output, f_trg) in state_f.transitions:
+   *     for (g_input, g_output, g_trg) in state_g.transitions:
+   *       if g_input == f_output:
+   *         gf_trg = gf.add_state()
+   *         gf.add_transition(gf_src,  f_input, g_output, gf_trg)
+   *         queue.add(f_trg, g_trg, gf_trg)
+   *     gf_trg = gf.add_state()
+   *     gf.add_transition(gf_src, f_input, f_output, gf_trg)
+   *     queue.add(f_trg, g.init, gf_trg)
+   *
+   * with some added complications:
+   *
+   * 1. Since we can have loops, we need to keep track of which pairs
+   *    (f_src, g_src) we've seen, so we don't keep re-adding them. And
+   *    on seeing a previously seen pair, we need to know which state
+   *    in gf was added from that pair.
+   *
+   * 2. We have to be able to skip epsilons on the input-side of g and
+   *    the output-side of f.
+   *
+   * 3. When f_inverted, we swap f_left/f_right in matching.
+   *
+   * 4. If composing g_anywhere, we can add initials from g anywhere
+   *    in gf, and finals in gf can be initials in g.
+   */
+  joinFinals(epsilon_tag);
+
+  // (f_state, g_state)
+  typedef std::pair<int, int> SearchState;
+
+  // State numbers will differ in fXg transducers and gf:
+  Transducer gf;
+  std::map<SearchState, int> states_f_g_gf;
+
+  std::vector<SearchState> todo;
+  std::set<SearchState> seen;
+  SearchState current;
+  SearchState next = std::make_pair(initial, g.initial);
+  todo.push_back(next);
+  states_f_g_gf.insert(std::make_pair(next, gf.initial));
+
+  while(!todo.empty()) {
+    current = todo.back();
+    todo.pop_back();
+    seen.insert(current);
+    int f_src  = current.first,
+        g_src     = current.second;
+
+    if(states_f_g_gf.find(current) == states_f_g_gf.end()) {
+      std::cerr <<"Error: couldn't find "<<f_src<<","<<g_src<<" in state map"<< std::endl;
+      exit(EXIT_FAILURE);
+    }
+    int gf_src = states_f_g_gf[current];
+
+    // First loop through _epsilon_ transitions of g (input side)
+    for(const auto &g_trans_it : g.transitions.at(g_src)) {
+      int g_label = g_trans_it.first,
+          g_trg   = g_trans_it.second.first;
+      double g_wt = g_trans_it.second.second; // TODO: Should the weight be weight of g, or multiplied/added?
+      int32_t g_left = g_a.decode(g_label).first;
+
+      if(g_left == 0)
+      {
+        next = std::make_pair(f_src, g_trg);
+        if (seen.find(next) == seen.end()) {
+          todo.push_back(next);
+        }
+        if (states_f_g_gf.find(next) == states_f_g_gf.end()) {
+          states_f_g_gf.insert(std::make_pair(next, gf.newState()));
+        }
+        int gf_trg = states_f_g_gf[next];
+        gf.linkStates(gf_src,
+                      gf_trg,
+                      epsilon_tag,
```

Comment:
maybe I've missed something, but if `g` has `0:X`, shouldn't the transition in `gf` also be `0:X`?

Reply:
Doh! Copy-pasta from lt-trim, thanks for catching that.